### PR TITLE
Enable Ruby 3.1 support

### PR DESCRIPTION
**Pull request by aharpervc**
#### Todo

- [x] add Ruby 3.1 to the supported versions list
- [x] switch the CI base image to `cimg/ruby`
- [x] update `.ruby-version` to 3.1.0
- [x] find & fix errors with Ruby 3.1
  - [x] ...in such a way that doesn't break old Ruby versions either

#### Notes

That gem version comparison approach was recommended [here](https://metaredux.com/posts/2018/10/28/a-better-way-to-compare-versions-in-ruby.html), amusingly for the exact same situation as we have with the changing YAML call signature.

### CODE REVIEW

**Review by sixfeetover — APPROVED** [review]
Looks good to me, and I'm glad that method signature has been cleaned up.

File: lib/consult.rb [hunk 1]
```diff
@@ -24,7 +24,15 @@ def load(config_dir: nil, force_render: false, verbose: nil)
       root directory: config_dir
       yaml = root.join('config', 'consult.yml')
 
-      @all_config = yaml.exist? ? YAML.safe_load(ERB.new(yaml.read).result, [], [], true, symbolize_names: true).to_h : {}
+      @all_config = if yaml.exist?
+        if Gem::Version.new(YAML::VERSION) < Gem::Version.new('4.0')
+          YAML.safe_load(ERB.new(yaml.read).result, [], [], true, symbolize_names: true).to_h
```
Comment (ronnietaylor) [inline lib/consult.rb:29]: Non-blocking comments:
In this call signature what are the 2nd and 3rd args (`[]`)? `legacy_permitted_classes` & `legacy_permitted_symbols`?

Is the 4th arg (`true) for `aliases` and can it be a named arg for greater clarity?

My curiosity was triggered by reading [the source for Psych 3.3.0](https://github.com/ruby/psych/blob/0abce07b908fa0128223a3e11f2eeb4be0c191d6/lib/psych.rb#L329) which is what Ruby 2.7.3 is using in my local.

**Review by ronnietaylor — APPROVED** [review]

**Comment by ronnietaylor** [comment]
Overall these changes seem ok

File: lib/consult.rb (hunk 1 above)
Comment (aharpervc) [reply lib/consult.rb:29]: > In this call signature what are the 2nd and 3rd args ([])? legacy_permitted_classes & legacy_permitted_symbols?

I assume so 🤷‍♂️

> Is the 4th arg (true) for `aliases` and can it be a named arg for greater clarity?

Again, I assume so; the objective of this code was to make as close to a 100% safe change as possible for the existing code path, but then fix it in the future code path. At some point we can come back here and delete the old code (probably when we drop support for Ruby < 3.1)

File: lib/consult/cli.rb [hunk 2]
```diff
@@ -15,7 +15,7 @@ class CLI
 
     def parse(args = ARGV)
       @opts = parse_options(args)
-      Consult.load @opts
+      Consult.load **@opts
```
Comment (samantharossetti) [inline lib/consult/cli.rb:18]: Why is this change needed?

Comment (aharpervc) [reply lib/consult/cli.rb:18]: The former syntax was passing a hash value to a function defined with keyword arguments, and relied on Ruby to automatically destructure a hash argument to match a each keyword argument. This feature was deprecated starting with Ruby 2.7 and removed in Ruby 3. Details: https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/ Per this page, the way to keep the behavior is to explicitly destructure your hash value with `**`.

Another approach would be to rewrite this line as `Consult.load config_dir: @opts['config_dir'], force_render: @opts['force_render'], verbose: @opts['verbose']` but the destructure seems simpler.

**Review by samantharossetti — APPROVED** [review]
LGTM

**Review by bvogelzang — APPROVED** [review]
